homeassistant/components/light/tradfri.py

```diff
@@ -8,25 +8,34 @@
 import logging
 
 from homeassistant.core import callback
-from homeassistant.const import ATTR_BATTERY_LEVEL
+import homeassistant.util.color as color_util
 from homeassistant.components.light import (
-    ATTR_BRIGHTNESS, ATTR_COLOR_TEMP, ATTR_RGB_COLOR, ATTR_TRANSITION,
+    ATTR_BRIGHTNESS, ATTR_COLOR_TEMP, ATTR_XY_COLOR, ATTR_TRANSITION,
     SUPPORT_BRIGHTNESS, SUPPORT_TRANSITION, SUPPORT_COLOR_TEMP,
-    SUPPORT_RGB_COLOR, Light)
+    SUPPORT_XY_COLOR, SUPPORT_RGB_COLOR, ATTR_RGB_COLOR, Light)
 from homeassistant.components.light import \
     PLATFORM_SCHEMA as LIGHT_PLATFORM_SCHEMA
 from homeassistant.components.tradfri import KEY_GATEWAY, KEY_TRADFRI_GROUPS, \
     KEY_API
-from homeassistant.util import color as color_util
 
 _LOGGER = logging.getLogger(__name__)
 
+ATTR_TRANSITION_TIME = 'transition_time'
 DEPENDENCIES = ['tradfri']
 PLATFORM_SCHEMA = LIGHT_PLATFORM_SCHEMA
 IKEA = 'IKEA of Sweden'
 TRADFRI_LIGHT_MANAGER = 'Tradfri Light Manager'
 SUPPORTED_FEATURES = (SUPPORT_BRIGHTNESS | SUPPORT_TRANSITION)
-ALLOWED_TEMPERATURES = {IKEA}
+
+
+def normalize_xy(argx, argy, brightness=None):
+    """Normalise XY to Tradfri scaling."""
+    return (int(argx*65535+0.56), int(argy*65535+0.56))
+
+
+def denormalize_xy(argx, argy, brightness=None):
+    """Denormalise XY from Tradfri scaling."""
+    return (argx/65535-0.56, argy/65535-0.56)
```

Comment:
These functions are not inverse of each other.
a*b+c = d
(d-c)/b = a

Only thing is I don't know which of the two functions is correct (yet)

Reply:
Denormalize is wrong, there should be not subtraction. If that's removed, the color picker will get correct.

Reply:
but then still, what's the point in adding 0.56 to the result of `arg*65535` in the normalize function. It gets rounded to an integer anyway.

Reply:
@NovapaX it was added into pytradfri when it supported color conversion. I believe that code was referenced from an external source. In other words, I don't know the maths behind 0.56.